### ISSUE TITLE
Error reading Parquet files from S3 with a space in the path.

### ISSUE THREAD

**Issue by pmleveque** [roapi/roapi#98]
Hi, I had a problem loading parquet file from s3 when there's a space in the path. I tried with `%20` but it doesn't work either.

Example path:  
s3://my-bucket/trusted/receita_socios/version=2021-10-30 09:00:56/my-file.parquet

Error message:
```
Error: Error loading Parquet: no parquet file found
```

The same file without space in the path works fine.

**Comment by houqp** [comment]
hmm, I am not able to reproduce this, uri with space results in `Error: Invalid table URI` error, which is expected. Changing it to `%20` works for me with aws localstack. Sorry I don't have real AWS s3 bucket setup for testing. Could you try running it with `RUST_LOG=debug`?

**Comment by PsypherPunk** [comment]
Ran into a similar issue and I can replicate this using a real S3 bucket.

If you attempt to pass through the file's location _with_ spaces:

```sh
--table "test=s3://${S3_BUCKET}/folder with spaces/data.parquet"
```

…then `roapi` warns you:

```sh
Error: Invalid table URI:…Make sure it's URI encoded.
```

If you then URI-encode it:

```sh
--table "test=s3://${S3_BUCKET}/folder%20with%20spaces/data.parquet"
```

…then `roapi` parses the URI correctly but you run into the `no parquet file found` error.

Using `RUST_LOG=debug`, I _think_ I can see what's happening:

```sh
[2022-10-03T09:16:33Z DEBUG rusoto_core::request] Full request: 
     method: GET
     final_uri: …?list-type=2&prefix=folder%2520with%2520spaces%2Fdata.parquet%2F&start-after=folder%2520with%2520spaces%2Fdata.parquet%2F
    Headers:
```

The spaces, encoded as `%20`, are getting doubly-encoded by the time `rusoto_core::request` makes its call.

I've verified this locally by having `columnq::io::s3::parse_uri()` URL-decode the `key` and it then works as expected. However, I'm not sure if that's necessarily desired behaviour…?

**Comment by jychen7** [comment]
we recently replace `rusoto` with `object_store` at https://github.com/roapi/roapi/issues/240, so I just try to reproduce this at https://github.com/jychen7/roapi/pull/8.

Interestingly, though both `use_memory_table: true` and `use_memory_table: false` use same `object_store` to read file now, the result is different.

Assume we have `blogs space.parquet`.

when `use_memory_table: true`
1. ❌ `blogs space.parquet`: Invalid table URI
2. ❌ `blogs%20space.parquet`: no parquet file found

when `use_memory_table: false`
1. ✅ `blogs space.parquet`: can query
2. ✅ `blogs%20space.parquet`: can query

I would need to investigate more...

**Comment by houqp** [comment]
@jychen7 I believe the first error came from `TableSource::parsed_uri`, we need to encode the path before parsing it there.

The 2nd error came from object store, it's `Path::from` implementation expects unencoded path, so it does its own encoding. We should switch to the parse method instead: https://docs.rs/object_store/latest/object_store/path/struct.Path.html#parse

**Comment by jychen7** [comment]
@houqp thank you. Then I think `blogs space.parquet` is expected to fail and we need to fix for `blogs%20space.parquet` first. I understand that `Path::from` below will double encode the `%20`. However, I try switch to `Path::parse`, still "no parquet file found", because object_store crate will not decode `%20` back to space.

https://github.com/roapi/roapi/blob/97e6a4cd8dde6ebf095ace08bfc7ae5943269c16/columnq/src/io/object_store.rs#L62

After look at Datafusion [ListingTableUrl](https://github.com/apache/arrow-datafusion/blob/master/datafusion/core/src/datasource/listing/url.rs), I now understand why both works in `use_memory_table: false`. `ListingTableUrl` use both `Url::parse` to encode possible char and `percent_decode` before calling `object_store` client

| input | Url::parse | percent_decode_str |
|-|-|-|
| blogs space.parquet | blogs%20space.parquet  | blogs space.parquet |
| blogs%20space.parquet | blogs%20space.parquet  | blogs space.parquet |

I adopt the percent_decode approach and fixed `blogs%20space.parquet` case at https://github.com/roapi/roapi/pull/257 first.

---
I think it is also good to use `Url::parse` before `URIReference::try_from`, so it is easier for user. What do you think ❓ @houqp . We can implement this in new PR if need.
https://github.com/roapi/roapi/blob/97e6a4cd8dde6ebf095ace08bfc7ae5943269c16/columnq/src/table/mod.rs#L408-L410

**Comment by houqp** [comment]
Thank you @jychen7 for the deepdive. I think the workarounds you proposed look fine as a quick fix to the problem :+1: So let's get them in.

Ideally, we should only need to encode/decode the path once. Right now, we are decoding the path, then pass the path to objectstore to reencode it, which seems like a waste of compute. It would be nice if we can change object store to accept an encoded path directly in the future.

**Comment by jychen7** [comment]
> It would be nice if we can change object store to accept an encoded path directly in the future.

agree, propose one idea at https://github.com/apache/arrow-rs/issues/3651

**Comment by jychen7** [comment]
I think it is also good to support `blogs space.parquet` in roapi config yml, so it is easier for user. 

one idea is to use Url::parse before URIReference::try_from, trade off is additonal encode operation (not ideal).

What do you think ❓ @houqp

**Comment by houqp** [comment]
@jychen7 I think the extra overhead here is not too bad since we only need to do it once per table during boot time. The extra encode/decode in the object store backend is what might have a small runtime performance impact because a table scan could touch hundreds or even thousands of objects.

**Comment by jychen7** [comment]
> one idea is to use Url::parse before URIReference::try_from

I just try at https://github.com/jychen7/roapi/pull/9/files#r1096708620, but still research how to fix `borrower`

**Comment by houqp** [comment]
@jychen7 thanks for taking a stab at this, I left my thoughts for the problem you are facing in a comment https://github.com/jychen7/roapi/pull/9/files#r1096807277.
